### PR TITLE
Add cached nvFuser's fusion creation for torch._prims.executor

### PR DESCRIPTION
**Pull request by IvanYashchuk**
In the current setup for each call of the `execute` function, a `Fusion` object was constructed using `GraphModule` and args, that's expensive.

This PR makes use of `functools.lru_cache` to pay the `Fusion` creation cost once per `GraphModule` and set of args. Currently, the shape, strides, and dtype of tensors are static it can be changed later to make better use of the nvFuser's internal caching mechanism (by specifying only ndim, contiguity, dtype).

On master:
```py
In [2]: a = torch.randn(3, 3, device='cuda')

In [3]: with TorchRefsMode.push():
   ...:     gm = make_fx(lambda x: torch.sigmoid(x))(a)
   ...:

In [4]: %%timeit
   ...: execute(gm, a, executor="nvfuser")
   ...: torch.cuda.synchronize()
175 ms ± 1.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
This PR:
```py
In [2]: a = torch.randn(3, 3, device='cuda')

In [3]: with TorchRefsMode.push():
   ...:     gm = make_fx(lambda x: torch.sigmoid(x))(a)
   ...:

In [4]: %%timeit
   ...: execute(gm, a, executor="nvfuser")
   ...: torch.cuda.synchronize()
62.6 µs ± 9.99 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

In addition, this PR adds support for pytree inputs and extends the test for this.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -258,6 +258,13 @@ void initNvFuserPythonBindings(PyObject* module) {
             return IrBuilder::create<Double>(val);
           },
           py::return_value_policy::reference)
+      .def(
```
Comment (IvanYashchuk) [inline torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:261]: These changes will be gone once https://github.com/pytorch/pytorch/pull/80522 is merged.

File: torch/_prims/executor.py [hunk 2]
```diff
@@ -21,57 +14,9 @@ def execute(gm: GraphModule, *args, executor: str = "aten"):
...
-                    x = fd.define_tensor(
-                        arg.size(), arg.stride(), getnvFuserDtype(arg.dtype)
-                    )
-                    fd.add_input(x)
-                    return x
-                else:
-                    return arg
-
-            # Transforms graph to call nvfuser lowerings
-            # Note, this doesn't handle nested structures in the args, TODO: add tree_flatten
```
Comment (IvanYashchuk) [inline torch/_prims/executor.py:61]: Now it works and is tested with `test_pytree_input_output`.

**Comment by IvanYashchuk** [comment]
I wanted to modify `nvFuserTensorViewTemplate` to store only `tensor.ndim` and `tensor.dtype`, but some tests started to fail due to https://github.com/csarofeen/pytorch/issues/1788, so let's leave this for later.

File: torch/_prims/nvfuser_executor.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+else:
+    DataType = None
+
+
+# nvFuserTensorViewTemplate and nvFuserScalarValTemplate are helper objects
+# for cached construction of the nvFuser's Fusion
+# TODO: change what is stored in the cache for nvFuser's TensorView objects
+# https://github.com/pytorch/pytorch/issues/80551
+@dataclass(frozen=True)
+class nvFuserTensorViewTemplate:
```
Comment (kevinstephano) [inline torch/_prims/nvfuser_executor.py:22]: I would remove the word `View` from this name.  I am going to remove direct exposure of the Fusion IR components like `TensorView` from the executor.

File: torch/_prims/nvfuser_executor.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    size: tuple
+    stride: tuple
+    dtype: DataType
+
+
+@dataclass(frozen=True)
+class nvFuserScalarValTemplate:
```
Comment (kevinstephano) [inline torch/_prims/nvfuser_executor.py:29]: Similarly, I would remove the world `Val` from this name to remove something that references a component of the Fusion IR.

File: torch/_prims/utils.py [hunk 5]
```diff
@@ -23,12 +23,17 @@
...
+        float: DataType.Double,
+        int: DataType.Int,
+        bool: DataType.Bool,
     }
 else:
     _torch_dtype_to_nvfuser_dtype_map = {}
 
 
-def getnvFuserDtype(dtype: torch.dtype):
+def getnvFuserDtype(dtype: Union[torch.dtype, type]):
```
Comment (kevinstephano) [inline torch/_prims/utils.py:36]: Question: Why is this a union?  I am not sure what the other `type` is.

**Review by kevinstephano — CHANGES_REQUESTED** [review]
I just wanted to ask for 2 function name changes to reflect up coming changes to the NVFuser Python interface then I think this should be good to go.

File: torch/_prims/utils.py (hunk 5 above)
Comment (IvanYashchuk) [reply torch/_prims/utils.py:36]: This function is now used to get the nvFuser's type for Python scalars (float, complex, bool, int) and the type of float/complex/bool/int is `type`.

Comment (kevinstephano) [reply torch/_prims/utils.py:36]: Thanks!

**Review by kevinstephano — APPROVED** [review]
Looks good to me.

File: torch/_prims/nvfuser_executor.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+# nvFuserTensorTemplate and nvFuserScalarTemplate are helper objects
+# for cached construction of the nvFuser's Fusion
+# TODO: change what is stored in the cache for nvFuser's Tensor objects
+# https://github.com/pytorch/pytorch/issues/80551
+@dataclass(frozen=True)
+class nvFuserTensorTemplate:
+    size: tuple
+    stride: tuple
+    dtype: DataType
```
Comment (jjsjann123) [inline torch/_prims/nvfuser_executor.py:25]: nitpitck: device might also be relevant here. Thinking about cpu scalar (today) and multi-device support (future-ish)

**Review by jjsjann123 — APPROVED** [review]
LGTM

File: torch/_prims/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply torch/_prims/utils.py:36]: Using `NumberType` is incorrect, mypy would complain. We're passing here not scalar, but their type, and in the type annotation we need to use the type of a type (`type(float) -> type`). But yeah `type` is too general and in Python 3.9+ it's possible to use `type[float]`. I understand that the type annotation in this case is not very helpful, I will change it to be more specific `Union[Type[float], Type[int], Type[bool], Type[complex]]`

File: torch/_prims/nvfuser_executor.py (hunk 6 above)
Comment (IvanYashchuk) [reply torch/_prims/nvfuser_executor.py:25]: What operations support cpu scalars? I tried `add(a, b)` and hit this
https://github.com/pytorch/pytorch/blob/76cff182428fbd165b5725f3de29dbd91a1512fa/torch/csrc/jit/codegen/cuda/executor_utils.cpp#L126-L134

Comment (jjsjann123) [reply torch/_prims/nvfuser_executor.py:25]: We should generally support cpu scalars, for `add` as well.
https://github.com/pytorch/pytorch/blob/76cff182428fbd165b5725f3de29dbd91a1512fa/test/test_jit_cuda_fuser.py#L4220-L4263

File: torch/_prims/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jjsjann123) [reply torch/_prims/utils.py]: Ha, sorry I was mistaken. :face_with_spiral_eyes:

**Review by SherlockNoMad — APPROVED** [review]

**Pull request closed without merging**